pkg/controller/labelidentity/label_group_index.go

```diff
@@ -0,0 +1,510 @@
+				// We can simply read matched label identity IDs from the index.
+				selectedLabelIDs := i.GetLabelIdentityIDs(selKey)
+				labelIdentityIDs = append(labelIdentityIDs, selectedLabelIDs...)
+				matchedSelectorKeys = append(matchedSelectorKeys, selKey)
+			}
+		}
+		// Remove matched clusterset-scoped selectors of the policy before and after the update.
+		// The selectors remaining in newSelectors will be new ones to be registered for the policy
+		// The selectors remaining in originalSelectors will need to be removed from the labelIdentityIndex
+		for _, k := range matchedSelectorKeys {
+			delete(originalSelectors, k)
+			delete(newSelectors, k)
+		}
+	}
+	for _, sel := range newSelectors {
+		selectedLabelIDs := i.AddSelector(sel, policyKey)
+		labelIdentityIDs = append(labelIdentityIDs, selectedLabelIDs...)
+	}
+	for selectorKey := range originalSelectors {
+		i.DeleteSelector(selectorKey, policyKey)
+	}
+	// Dedup labelIdentityIDs in-place
+	seen := map[uint32]struct{}{}
+	idx := 0
+	for _, id := range labelIdentityIDs {
+		if _, exists := seen[id]; !exists {
+			seen[id] = struct{}{}
+			labelIdentityIDs[idx] = id
+			idx++
+		}
+	}
+	return labelIdentityIDs[:idx]
+}
+
+func (i *LabelIdentityIndex) getPolicySelectors(policyKey string) map[string]*types.GroupSelector {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	res := map[string]*types.GroupSelector{}
+	selectors, _ := i.selectorItems.ByIndex(policyIndex, policyKey)
+	for _, s := range selectors {
+		sel := s.(*selectorItem)
+		res[sel.getKey()] = sel.selector
+	}
+	return res
+}
+
+func (i *LabelIdentityIndex) DeletePolicySelectors(policyKey string) {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	selectors, _ := i.selectorItems.ByIndex(policyIndex, policyKey)
+	for _, s := range selectors {
+		sel := s.(*selectorItem)
+		i.deleteSelector(sel.getKey(), policyKey)
+	}
+}
+
+func (i *LabelIdentityIndex) GetLabelIdentityIDs(selectorKey string) []uint32 {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	if s, exists, _ := i.selectorItems.GetByKey(selectorKey); exists {
+		sel := s.(*selectorItem)
+		return i.getMatchedLabelIdentityIDs(sel)
+	}
+	return []uint32{}
+}
+
+func (i *LabelIdentityIndex) getMatchedLabelIdentityIDs(sItem *selectorItem) []uint32 {
+	var ids []uint32
+	for lKey := range sItem.labelIdentityKeys {
+		labelIdentity := i.labelIdentities[lKey]
+		ids = append(ids, labelIdentity.id)
+	}
+	return ids
+}
+
+func (i *LabelIdentityIndex) scanLabelIdentityMatches(labelIdentityKeys sets.String, sItem *selectorItem) {
+	for lkey := range labelIdentityKeys {
+		labelIdentity := i.labelIdentities[lkey]
+		if labelIdentity.matches(sItem) {
+			sItem.labelIdentityKeys.Insert(lkey)
+			labelIdentity.selectorItemKeys.Insert(sItem.getKey())
+		}
+	}
+}
+
+func (i *LabelIdentityIndex) AddLabelIdentity(labelKey string, id uint32) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	existingLabelMatch, exists := i.labelIdentities[labelKey]
+	if exists {
+		if existingLabelMatch.id != id {
```

Comment:
It should not happen. But it will make sure that we can recover from the case where a LabelIdentity delete is not handled correctly, while later the same label identity is added back with a different ID